### DISCUSSION TITLE
Avoid BEGIN/COMMIT around INSERT

### DISCUSSION THREAD

**Discussion by boje** [sqlalchemy/sqlalchemy#8870] [Usage Questions]
We have been testing SQLAlchemy for our integration to Snowflake. 
When we did a load test, we experienced it to be quite slow. This is because Snowflake block writes into a table until a COMMIT is done and a BEGIN/COMMIT is slow in Snowflake.

Example:
When the code below is executed simultaneously 
```
person = Person(username=<username_value>, email=<email_value>)
db.session.add(person)
db.session.commit()
```
it ends out in Snowflake as
```
BEGIN 
INSERT INTO members          ("username",   "email")
VALUES ('X, '47b0fcf9@test.com')
COMMIT  

```

*Questions:*
Is there any way that `db.session.add(person)` can send without `BEGIN` and `COMMIT`?
Or is there a way to extract the SQL code generated before it is sent to Snowflake?

**Comment by zzzeek** [comment]
yes, you would make use of Snowflake's "Autocommit" mode, assuming it has one available.     See https://docs.sqlalchemy.org/en/14/core/connections.html#understanding-the-dbapi-level-autocommit-isolation-level

**Reply by zzzeek** [reply]
here's ORM specific examples  https://docs.sqlalchemy.org/en/14/orm/session_transaction.html#session-transaction-isolation

**Reply by boje** [reply]
If we add `isolation_level="AUTOCOMMIT"` we get this error:
```
TypeError: Invalid argument(s) 'isolation_level' sent to create_engine(), using configuration SnowflakeDialect/QueuePool/Engine.  Please check that the keyword arguments are appropriate for this combination of components.

File "/home/datarise/projects/bestseller/dna-b2c-customer-api/.venv/lib/python3.9/site-packages/sqlalchemy/engine/interfaces.py", line 942, in set_isolation_level    raise NotImplementedError()
NotImplementedError
```
[Here](https://docs.snowflake.com/en/user-guide/sqlalchemy.html#opening-and-closing-a-connection) it shows how to set AUTOCOMMIT in SQLAlchemy
```
# Disable AUTOCOMMIT if you need to use an explicit transaction.
with engine.connect().execution_options(autocommit=False) as connection:
```
[More info](https://docs.snowflake.com/en/sql-reference/transactions.html#label-txn-autocommit)

If I look at [isolation level](https://docs.snowflake.com/en/sql-reference/transactions.html#isolation-level) it mention that only READ COMMITTED is supported.

I'm a bit confuse of `AUTOCOMMIT` have two menings here?

**Reply by datarise** [reply]
Hi, I hope it is okay that I add to this. 

Is it possible to enable the isolation level in another way if it is not implemented by the engine?

**Reply by zzzeek** [reply]
reading https://docs.snowflake.com/en/user-guide/python-connector-example.html

it looks like snowflake connector does not have an "autocommit" attribute that can be set on an existing connection.   as such, it looks like the snowflake dialect does not implement SQLAlchemy's "autocommit" feature directly.

however, the connector appears to have an [autocommit parameter to connect](https://docs.snowflake.com/en/user-guide/python-connector-example.html#using-context-manager-to-connect-and-control-transactions).   so use that

```py
e = create_engine("snowflake+....://....", connect_args={"autocommit": True})
```

I would raise an issue at https://github.com/snowflakedb/snowflake-sqlalchemy that they should seek to support SQLAlchemy's "AUTOCOMMIT" feature fully, and in the absence of that should provide documentation as to a temporary workaround for now.

**Reply by wiseaidev** [reply]
Just wanted to chime in. I ran into a similar issue before. Basically, you have to set the `autocommit` mode on an engine instance like the following:

```python
autocommit_engine = engine.execution_options(isolation_level="AUTOCOMMIT")
```

You can refer to [this module](https://github.com/brave-chat/brave-chat-server/blob/5b91ba6a0cb284256ea541e77f8e1d3b5b10754c/app/utils/engine.py#L85) for more info. Hope it helps.
